universe/syncer.go

```diff
@@ -236,14 +264,59 @@ func (s *SimpleSyncer) syncRoot(ctx context.Context, remoteRoot Root,
 		uniID.String())
 
 	// Otherwise, we'll need to perform a diff operation to find the set of
-	// keys we need to fetch.
-	remoteUniKeys, err := diffEngine.UniverseLeafKeys(ctx, uniID)
-	if err != nil {
-		return err
+	// keys we need to fetch. We'll start by fetching the set of keys from
+	// both the local and remote Universe.
+	var (
+		remoteUniKeys []LeafKey
+		localUniKeys  []LeafKey
+	)
+
+	// Initialize the offset to be used for the pages.
+	offset := int32(0)
+
+	for {
+		tempRemoteKeys, err := diffEngine.UniverseLeafKeys(
+			ctx, UniverseLeafKeysQuery{
+				Id:            uniID,
+				Offset:        offset,
+				Limit:         defaultPageLimit,
+				SortDirection: SortAscending,
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		if len(tempRemoteKeys) == 0 {
+			break
+		}
+
+		remoteUniKeys = append(remoteUniKeys, tempRemoteKeys...)
+		offset += defaultPageLimit
 	}
-	localUniKeys, err := s.cfg.LocalDiffEngine.UniverseLeafKeys(ctx, uniID)
-	if err != nil {
-		return err
+
+	// Initialize the offset to be used for the pages.
+	offset = 0
+
+	for {
+		tempLocalKeys, err := s.cfg.LocalDiffEngine.UniverseLeafKeys(
```

Comment:
Same here re extracting into a `fetchAllLeafkeys` function. Can eliminate the duplication above.